### PR TITLE
ENH: enable approximation for factorialk 

### PR DESCRIPTION
**Pull request by h-vetinari**
Follow-up to https://github.com/scipy/scipy/pull/15841, preparation for further work towards https://github.com/scipy/scipy/issues/18409 (otherwise the API for factorialk would be very strange), and preparing to make `exact=` have the same default across all factorial* functions.

Based on #19703 and a draft until that's merged (although I can separate the PRs if desired), until then only the last commit is relevant.

### CODE REVIEW

**Comment by h-vetinari** [comment]
```
scipy/special/_basic.py:3149:9: E731 Do not assign a `lambda` expression, use a `def`
scipy/special/tests/test_basic.py:2463:9: E731 Do not assign a `lambda` expression, use a `def`
```
strongly disagree in these two cases

**Comment by lucascolley** [comment]
You can add `  # noqa: E731` to the lines (unless you mean you want to propose we remove that rule project-wide)

**Comment by h-vetinari** [comment]
Now rebased after #19703 went in; failure is unrelated (#19737), PTAL @steppi @j-bowhay

**Comment by h-vetinari** [comment]
The last three commits could be split off into a separate PR, but I feel they're very related here. I've pulled them forward from work towards making `factorial{,2,k}` support complex inputs.

**Comment by tovrstra** [comment]
I could not spot anything worth commenting on. I just wonder if this changes is already expected to produce `factorial2(-1)==1`? If so, it might be useful including it in the unit tests. Thanks for the nice work!

**Comment by h-vetinari** [comment]
> I just wonder if this changes is already expected to produce `factorial2(-1)==1`? If so, it might be useful including it in the unit tests.

Nope, the current behaviour stays the same for all the factorial functions, which is that `f(x) = 0` for all `x<0`. This is the historic behaviour and will be named `extend='zero'` (the default, obviously) once the support for `extend='complex'` lands.

I'm not sure which behaviour you have in mind specifically for `factorial2(-1)==1`, because there are two possible reasons. The one - which is the case - is that the complex extension _by accident_ happens to have `factorial2(-1, extend='complex')==1`, because of the way various factors in the formula cancel out.

The other reason - which is not and won't be the case - is that it's possible to define the (multi-)factorial in a way such that the recursion stops only for `n ∈ [-k,-1]` with ones for those negative integers. For lack of a better word, that's only a mathematical implementation detail though, and it's completely possible to define the multifactorial consistently in a way that the recursion stops for `n ∈ [0, k-1]`:
```
n!(k) = n * (n - k)!(k) if n ≥ k
      = n if 0 ≤ n < k
      = 0 if n < 0
```

This is a much saner choice than `extend='zero-except-ones-at-negative-integers-up-to-minus-k'`, which is why we went with that.

**Comment by h-vetinari** [comment]
Gentle ping @steppi 

If you're unconcerned/indifferent about this PR, I'm happy enough to merge it as-is, but wanted to wait for your input first.

**Comment by steppi** [comment]
Hey @h-vetinari, I’ve had this on my radar, but have been on vacation recharging for a bit. I‘ll give it a close look in the next couple weeks.

**Review by j-bowhay — COMMENTED** [review]
Thanks @h-vetinari, does removing the special casing of factorial2 exact=False lose any performance?

File: scipy/special/_basic.py
Comment (h-vetinari) [reply scipy/special/_basic.py:2902]: There's a comment directly below?

File: scipy/special/_basic.py
Comment (h-vetinari) [reply scipy/special/_basic.py:2915]: That's actually a pretty good catch! But we need to exclude 0.

File: scipy/special/_basic.py
Comment (j-bowhay) [reply scipy/special/_basic.py:2902]: Sorry, what I meant was, when I saw the scalar branch comment I expected to see `np.scalar` but having read the issue maybe that's just an anti-pattern I need to erase from my brain.

**Comment by h-vetinari** [comment]
> does removing the special casing of factorial2 exact=False lose any performance?

You mean because the same computation (previously in `_approx`) is now done in a more general function?

Aside from trivial overhead like `if k == 1` and `if not isinstance(n, np.ndarray)`, the computation is exactly the same. I do not want two almost-identical implementations between factorial2 & factorialk, that's a permanent maintenance headache not worth some micro-optimizations.

File: scipy/special/_basic.py
Comment (h-vetinari) [reply scipy/special/_basic.py:2902]: OK, I get your point now; I'll think about how I document this so the comment isn't duplicated X times.

Comment (h-vetinari) [reply scipy/special/_basic.py:2902]: PS. Some of this will change once the complex extensions land; I have a branch that unifies all the underlying implementations of `factorial{,2,k}`. I'm just waiting for this PR to land before I open the next PR.

File: scipy/special/_basic.py
Comment (h-vetinari) [reply scipy/special/_basic.py]: It would be an iota more efficient to write `set(np.unique(n_mod_k)) - {0}`, but for now I went with the more readable `if r == 0: continue`.

**Comment by h-vetinari** [comment]
> My use case is probably for the `zero-except-ones-at-negative-integers-up-to-minus-k` scenario. This happens to be convenient for the implementation of solutions of recursion relations.

If it's really _that_ much more convenient, I suggest you create your own wrapper:
```python
from scipy.special import factorialk
def my_factorialk(n, k):
    res = factorialk(n, k)
    if isinstance(n, np.ndarray):
        res[(n > -k) & (n < 0)] = 1
    else:
        res = 1 if (n > -k) and (n < 0) else res
    return res
```

Otherwise, if you can come up with a catchy name for the extension you'd like to have (that doesn't conflict with the _other_ recurrence-based extension for `factorial2`), please make a comment in https://github.com/scipy/scipy/issues/18409. Perhaps you can convince people that it's worth supporting this special case.

**Review by j-bowhay — APPROVED** [review]
This now all seems good to me. I will leave for a bit as it looked like @steppi was planning to take a look. As non-blocking comments, I would be in favour of adding a reference to https://en.wikipedia.org/wiki/Double_factorial#Alternative_extension_of_the_multifactorial and an `exact=False` line in the example

**Comment by tovrstra** [comment]
@h-vetinari Thanks for your helpful comments. I'll keep it in mind but I don't want to stand in the way of the ongoing improvements. It's not that critical.

**Review by steppi — COMMENTED** [review]
Nice work. Besides some very small suggestions, I had one question. You seem to have gone out of your way a bit to avoid writing a scalar kernel for the approximation. Any particular reason?

File: scipy/special/_basic.py
Comment (h-vetinari) [reply scipy/special/_basic.py:3107]: I can do that, though the terms mean the same thing (at least etymologically). In my math education, the nomenclature w.r.t. equivalence relations was always "residue" / "residual" (though "remainder" is common when used with division explicitly).

File: scipy/special/_basic.py
Comment (h-vetinari) [reply scipy/special/_basic.py:3110]: There's a short proof in the commit message, which seemed like an appropriate place (I felt it would be out-of-place in the docstring).

File: scipy/special/_basic.py
Comment (h-vetinari) [reply scipy/special/_basic.py:3105]: Fine by me.

File: scipy/special/_basic.py
Comment (h-vetinari) [reply scipy/special/_basic.py:2903]: Not sure we documented the expectation anywhere, except in tests of course resp. the discussions for #15600 which eventually led to #18768.

Do you want to document this in the docstring?

File: scipy/special/_basic.py
Comment (h-vetinari) [reply scipy/special/_basic.py:2920]: With "control flow", you mean the application of the correction factors here?

**Review by h-vetinari — COMMENTED** [review]
Thanks for the review!

> You seem to have gone out of your way a bit to avoid writing a scalar kernel for the approximation. Any particular reason?

No reason that rises to the standard of "it must be so", but the reason why I wrote things as they are:
* the factorial functions are currently implemented in pure python; adding `approx=False` for `factorialk` did not seem a strong enough reason to change that.
* At the same time, the implementation would be duplicated between `factorial2` & `factorialk`, so I unified that.
* I didn't want to have to investigate and fix (much less consider a deprecation cycle for) existing differences between `special.factorial` and the `_ufuncs._factorial` (e.g. the ufunc machinery hasn't yet adopted #18768, whereas the factorial functions have).

I agree that these things are worthwhile improvements, but they're IMO orthogonal to this PR, and should be handled separately. I can also post the rest of the pending factorial work in a PR so you can see where I was planning to go with this - there's a bunch more feature work and unification coming up; perhaps that makes it easier to decide where/how to drop this down to the ufunc level.

In either case, I wouldn't want to couple the ufunc-question to this PR (though I can accept putting a hold on other changes until we've fixed this).

**Comment by steppi** [comment]
> I agree that these things are worthwhile improvements, but they're IMO orthogonal to this PR, and should be handled separately. I can also post the rest of the pending factorial work in a PR so you can see where I was planning to go with this - there's a bunch more feature work and unification coming up; perhaps that makes it easier to decide where/how to drop this down to the ufunc level.

I don't think they're quite orthogonal, but I don't think it's a reason to block this, as long as the behavior is as intended. So sounds good, we can wait until everything is settled before writing any compiled code.

File: scipy/special/_basic.py
Comment (steppi) [reply scipy/special/_basic.py:3107]: Yeah, I know they mean the same thing. I'm just thinking in terms of accessibility. I didn't encounter residue used this way in a class until the 3rd year of my math undergraduate program, but I've known what a remainder is since primary school.

File: scipy/special/_basic.py
Comment (steppi) [reply scipy/special/_basic.py:3110]: Yeah, I don't think it's needed in the docstring either.

File: scipy/special/_basic.py
Comment (steppi) [reply scipy/special/_basic.py:2903]: Yeah. I think we should add a note like this in the docstring of all the combinatorial functions which preserve 0d-arrays.

File: scipy/special/_basic.py
Comment (steppi) [reply scipy/special/_basic.py:2920]: Yeah, I mean that you're applying the correction factor based on conditional logic. It took me a minute to understand what was going on here, but a simple scalar kernel I would have grasped in a second. If you'll be able to work on this more productively using pure python up until the point where things are settled, I think it's fine to keep it like this for now.

**Review by steppi — APPROVED** [review]
OK, modulo moving the description of the approximation used to the Notes section and adding a note that the function preserves 0d-arrays, I think this is ready.

**Comment by h-vetinari** [comment]
@steppi @j-bowhay 

I think I've now incorporated all the feedback (minus the ufunc-question, which I'm glad we can defer; I'll open a follow-up issue after merging). Please feel free to take another look if you want, otherwise I'm planning to merge this on Monday.

File: scipy/special/_basic.py
Comment (steppi) [inline scipy/special/_basic.py:3157]: I think we should mention what the default behavior is, and the motivation behind breaking the convention for combinatorial functions. Maybe saomething like

The default behavior for [ufuncs](https://numpy.org/doc/stable/reference/ufuncs.html) is to silently convert the output to scalar for 0-dimensional array input but `factorialk` and other combinatorial functions in SciPy will preserve shape when given  0-dimensional input. This is because in some cases they will return object arrays of arbitrary sized integers, and conversion to scalar will result in the output being a Python ``int`` object which would be missing the array attributes and methods which are present for NumPy scalars.

Comment (h-vetinari) [reply scipy/special/_basic.py:3157]: I'm not sure I'd want to call it "default behaviour" as much as "inherited legacy behaviour". There's no good a priori reason for converting the type, and in the RFC there seemed to be broad agreement that this is more historical artefact rather than intentional.

I'll reformulate the note as you wish, but I feel this is getting lost too much in the weeds to be useful for a docstring TBH.

Comment (steppi) [reply scipy/special/_basic.py:3157]: I’d say don’t reformulate it if you don’t agree. I meant it as a suggestion, and wouldn’t want you to make the change without consensus. At the moment, I think the one sentence seems lacking in context, but maybe the ideal thing would be to link somewhere else that has more detailed information, leaving the details out of the docstring. Honestly; I think it’s fine to punt on documenting this behavior for now and leave mention of this behavior for a future PR. That way we can avoid holding this up without forcing through something suboptimal.

Comment (h-vetinari) [reply scipy/special/_basic.py]: OK, I removed the last commit for now, it seems we should also deal with this question independently from this PR. :)

**Comment by h-vetinari** [comment]
Test failure is unrelated, merging

**Pull request merged**
